src/Components/Server/src/Circuits/CircuitSecret.cs

```diff
@@ -0,0 +1,38 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    // Using a type to represent circuit secrets so we can't mistake them
+    // for other kinds of strings/ids.
+    //
+    // Intentionally not overriding ToString here so that this won't accidentally
+    // get logged. It's ok to log the secret at TRACE.
+    internal readonly struct CircuitSecret : IEquatable<CircuitSecret>
+    {
+        public CircuitSecret(string secret)
+        {
+            Secret = secret ?? throw new ArgumentNullException(nameof(secret));
+        }
+
+        public string Secret { get; }
+
+        public bool Equals([AllowNull] CircuitSecret other)
+        {
+            return string.Equals(Secret, other.Secret, StringComparison.Ordinal);
+        }
+
+        public override bool Equals(object obj)
+        {
+            return obj is CircuitSecret other ? Equals(other) : false;
+        }
+
+        public override int GetHashCode()
+        {
+            return Secret?.GetHashCode() ?? 0;
```

Comment:
Id rather us do `HashCode.Combine(Secret)` even if in this case is very similar. The performance gain if any here should be negligible. And `HashCode.Combine` is the standard approach.